compiler/rustc_lint/src/levels.rs

```diff
@@ -510,27 +526,34 @@ impl<'s> LintLevelsBuilder<'s> {
             }
         }
 
-        let prev = self.cur;
-        if !specs.is_empty() {
-            self.cur = self.sets.list.push(LintSet { specs, parent: prev });
+        if self.current_specs().is_empty() {
+            self.sets.list.pop();
+            self.cur = prev;
         }
 
         BuilderPush { prev, changed: prev != self.cur }
     }
 
     /// Checks if the lint is gated on a feature that is not enabled.
-    fn check_gated_lint(&self, lint_id: LintId, span: Span) {
+    ///
+    /// Returns `true` if the lint's feature is enabled.
+    fn check_gated_lint(&self, lint_id: LintId, span: Span) -> bool {
         if let Some(feature) = lint_id.lint.feature_gate {
             if !self.sess.features_untracked().enabled(feature) {
-                feature_err(
-                    &self.sess.parse_sess,
-                    feature,
-                    span,
-                    &format!("the `{}` lint is unstable", lint_id.lint.name_lower()),
-                )
-                .emit();
+                let (unknown_lints_level, _) = self.lint_level(builtin::UNKNOWN_LINTS);
+                if unknown_lints_level != Level::Allow {
+                    feature_err(
```

Comment:
would it make sense to reuse the `UNKNOWN_LINTS` lint here as well, adding a note that the lint is currently unstable and that this attribute does not enable it?

Reply:
I think this is a good idea, but this will cause a behavior change. Currently, using an unstable lint results in a hard error by default. This change would make using an unstable lint into a warning by default because the `unknown_lints` lint is set to `Warn` by default.

I put this together locally; here's what the output would look like:

`-Wunknown_lints` (default)

```
warning[E0658]: the `test_unstable_lint` lint is unstable
   |
   = help: add `#![feature(test_unstable_lint)]` to the crate attributes to enable
   = note: this is ignored but not enabled by the `unknown_lints` lint
```

`-Dunknown_lints`

```
error[E0658]: the `test_unstable_lint` lint is unstable
  |
  = help: add `#![feature(test_unstable_lint)]` to the crate attributes to enable
```

`-Aunknown_lints` ignores the lint and produces no output.